[red-knot] Move tuple-containing-Never tests to Markdown (#15402)
astral-sh/ruff: crates/red_knot_python_semantic/resources/mdtest/type_properties/tuples_containing_never.md

# Tuples containing `Never`

> A `tuple` that contains `Never` simplifies to `Never`. One way to think about this is the following:
> in order to construct a tuple, you need to have an object of every element type. But since there is
> no object of type `Never`, you can not construct a tuple that contains `Never`. The tuple type is
> uninhabited and therefore equivalent to `Never`.

Reviewer: I think this confuses the ideas of runtime tuples and tuple _types_ a little bit... what about something like this?

```suggestion
A tuple type that has `Never` as a type argument simplifies to `Never`. One way to think about this is the following. For any heteregenous tuple type `T`, `T` will have type arguments E<sub>1</sub>, E<sub>2</sub>, E<sub>3</sub>... where each type argument represents the type a runtime tuple element must have at a given index if the runtime tuple as a whole can be said to inhabit `T`. Therefore, to construct a runtime tuple `t` that inhabits `T`, it must be possible to construct a runtime object that inhabits E<sub>2</sub>, E<sub>3</sub>, a runtime object that inhabits E<sub>2</sub>, E<sub>3</sub>... But since no runtime objects can ever inhabit the type `Never`, if `Never` is present as one of the type arguments for `T` you can therefore not construct any inhabitants of `T` at runtime. `T` is therefore uninhabited, and therefore equivalent to `Never`. 
```

Author: Hm, to be honest, I'm not sure if that helps with clarity? What part of my description confused runtime tuples with tuple types? Maybe I should have written it as
```suggestion
A heterogeneous `tuple[…]` type that contains `Never` as a type argument simplifies to `Never`.
One way to think about this is the following: in order to construct a tuple, you need to have an
object of every element type. But since there is no object of type `Never`, you can not construct
the tuple. Such a tuple type is therefore uninhabited and equivalent to `Never`.
```

Reviewer: > What part of my description confused runtime tuples with tuple types? The word "contains"?

yes -- when you say "a `tuple` that contains `Never`, I think the natural understanding of `tuple` there is "a runtime tuple" rather than "a tuple type", and even for a tuple _type_ I wouldn't really say that it "contains `Never`"

> Maybe I should have written it as

that rewrite looks great! "can not" should be "cannot" here, though

> In the language of algebraic data types, a `tuple` is a product type and `Never` acts like the zero

Reviewer: ```suggestion
In the language of algebraic data types, a tuple type is a product type and `Never` acts like the zero
```
